Python/optimizer.c

```diff
@@ -0,0 +1,245 @@
+        int new_capacity = capacity ? capacity * 2 : 4;
+        _PyExecutorArray *new = PyMem_Realloc(
+            old,
+            offsetof(_PyExecutorArray, executors) +
+            new_capacity * sizeof(_PyExecutorObject *));
+        if (new == NULL) {
+            PyErr_NoMemory();
+            return -1;
+        }
+        new->capacity = new_capacity;
+        new->size = size;
+        code->co_executors = new;
+    }
+    assert(size < code->co_executors->capacity);
+    code->co_executors->size++;
+    return size;
+}
+
+static int
+insert_executor(PyCodeObject *code, int offset, _PyExecutorObject *executor)
+{
+    _Py_CODEUNIT original = _PyCode_CODE(code)[offset];
+    if (original.op.code == ENTER_EXECUTOR) {
+        _PyExecutorObject *old = code->co_executors->executors[original.op.arg];
+        executor->vm_data.opcode = old->vm_data.opcode;
+        executor->vm_data.oparg = old->vm_data.oparg;
+        old->vm_data.opcode = 0;
+        Py_INCREF(executor);
+        code->co_executors->executors[original.op.arg] = executor;
+        Py_DECREF(old);
+    }
+    else {
+        int32_t index = get_next_free_in_executor_array(code);
+        if (index < 0) {
+            return -1;
+        }
+        Py_INCREF(executor);
+        executor->vm_data.opcode = original.op.code;
+        executor->vm_data.oparg = original.op.arg;
+        code->co_executors->executors[index] = executor;
+        assert(index < 256);
+        _PyCode_CODE(code)[offset].op.code = ENTER_EXECUTOR;
+        _PyCode_CODE(code)[offset].op.arg = index;
+    }
+    return 0;
+}
+
+int
+PyUnstable_Replace_Executor(PyCodeObject *code, int offset, _PyExecutorObject *new)
+{
+
+    _Py_CODEUNIT original = _PyCode_CODE(code)[offset];
+    if (original.op.code != ENTER_EXECUTOR) {
+        PyErr_Format(PyExc_ValueError, "No executor to replace");
+        return -1;
+    }
+    return insert_executor(code, offset, new);
+}
+
+static _PyExecutorObject *
+error_optimize(
+    _PyOptimizerObject* self,
+    PyCodeObject *code,
+    _Py_CODEUNIT *instr)
+{
+    PyErr_Format(PyExc_SystemError, "Should never call error_optimize");
+    return NULL;
+}
+
+static PyTypeObject DefaultOptimizer_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "noop_optimizer",
+    .tp_basicsize = sizeof(_PyOptimizerObject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_DISALLOW_INSTANTIATION,
+};
+
+_PyOptimizerObject _PyOptimizer_Default = {
+    PyObject_HEAD_INIT(&DefaultOptimizer_Type)
+    .optimize = error_optimize,
+    .resume_threshold = UINT16_MAX,
+    .backedge_threshold = UINT16_MAX,
+};
+
+_PyOptimizerObject *
+PyUnstable_GetOptimizer(void)
+{
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    if (interp->optimizer == &_PyOptimizer_Default) {
+        return NULL;
+    }
+    Py_INCREF(interp->optimizer);
+    return interp->optimizer;
+}
+
+void
+PyUnstable_SetOptimizer(_PyOptimizerObject *optimizer)
```

Comment:
I thought at one point the idea was that we could have multiple optimizers installed, with different cost/benefit numbers, and the VM could decide which to use on that basis.

It seems fine to start with something simpler, but it seems like this `SetOptimizer/GetOptimizer` API kind of locks us into the "there is only one optimizer" world. Do we intend to stay there, or do we intend to maybe later support multiple optimizers, in which case we'll basically have to fully replace this API with something a little more involved?

Reply:
Just one for now. 
I suspect the only reason for multiple optimizers, is something like PyTorch Dynamo which wants to let the default optimizer handle code that it cannot handle. In that case the custom optimizer can call the default optimizer.

Reply:
This API is the interface between tiers 1 and 2. The executors would need their own mechanism to ramp up to tier 3, as the executor may never return to the tier 1 interpreter.